### ISSUE TITLE
Speed Ez - Account #5195

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#1757]
In GitLab by @kdjstudios on Oct 28, 2019, 14:32

**Submitted by:** Tammy Padalecki <tammy@speedez.com>

**Helpdesk:** http://www.servicedesk.answernet.com/profiles/ticket/2019-10-28-16663/conversation

**Server:** External

**Client/Site:** Speed Ez

**Account:** 5195

**Issue:** 

Is there any way possible that you would be able to either call our ofc and speak to Teresa Paige (210)615-0964 or you are welcome to email her at sales@speedez. We are having an issue with codes for Dispatching. Acct #5195 does not show any overages and they should have only 4000 for SMS, but our report shows 4578. We don’t see any overages at all. Could you help us to understand invoice SPDZ-29721?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 28, 2019, 14:44

I wrote:

> We took a quick look at the account 5195 and the usage that was uploaded to SA Billing. We have attached the filtered and highlighted copy of the usage file and it shows zeros for all of he metrics included in the 2051 Dispatching billing code. Please let us know if you have any questions or concerns. 

[resource_usage_09_01_09_30_-_5195_Highlighted.xlsx](/uploads/d361ba43841e553407a311d918481548/resource_usage_09_01_09_30_-_5195_Highlighted.xlsx)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 29, 2019, 08:48

Teresa wrote:

> When we run the usage report  on our end for September, we find that they had 1719 in SMS's and 2859 SMP which exceeds the combined total of  4000, which should have accumulated 578 in overages for the month but that wasn’t on the billing.  So is this something that is incorrect in billing codes that we may need to change.

I wrote:

> SA Billing can only calculate metrics from what is uploaded to via the usage file. The file that was uploaded did not have any usage in those metrics and I attached that file in my previous email. We would recommend that  you  look into your switch to ensure that it is exporting usage correctly. Please let us know if you have any questions or concerns. 

Teresa wrote:

> Per Mrs. Browne, on the human readable report, we see the corresponding numbers from our report to yours with the correct amount of text messages (SMS) , what is missing the count for SMP text messages.  So do we need to code for them?  


Jo Ann wrote:

> Ok guys I need 2054 (SMS+) attached to 2051 along with the other 5 that is already there. We were NOT collecting this data at all.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 29, 2019, 08:50

@smasih @amishra @jredfield 

May we please have an estimate on adding the 2054 code into the 2051 code. I took a quick look at the 2054 code on the HR report but it does not have any value setup to be importing to it by the looks of it; same with 2052 and 2053. So may we please get the user and date when those codes were created, so we can confirm if the client requested any formula to go with the code.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 29, 2019, 09:07

I wrote:

> We looked at the HR report and to me it would appear the Dispatching Services 2051 Billing code does already include SMP:Count. We have also attached a screenshot and included the current formula below too. Would you please be kind enough to clarify which column from the uploaded usage file you wish to add into this formula? Thank you and we look forward to your response. 
> 
> Dispatching Services 2051 = Alpha:Count + Digital:Count + Secure Msg:Count + SMS:Count + SMP:Count

I wrote:

> We took a look at the HR report and it would appear the the SMS+ 2054 billing code currently does not import any values from the uploaded usage, as you may see there is no formula or metrics in the last two column in the screen shot. It also looks like the the 2052 and 2053 billing codes do not import any usage either.  We are having the support team confirm who created these Billing codes and will follow up accordingly. 
> 
> In conjugation with the emails from Terasa from yesterday, we would like to confirm what the exact formula for the Dispatching Services 2051 billing code is desired to be. Please include the column headers labels for each metric that you wish to include from the usage file uploaded to SA Billing. We have included the current formula below. 
> 
> Dispatching Services 2051 = Alpha:Count + Digital:Count + Secure Msg:Count + SMS:Count + SMP:Count
> 
> Please let us know if there are any questions or concerns.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 29, 2019, 14:37

@smasih @amishra @jredfield 

After further review and discussion with Jo Ann and her team, it would appear that we are possibly not summing together all sub accounts. Please see the attached file that is being uploaded. 

Uploaded: [ResourceUsage_09-01_09-30.csv](/uploads/ca94509a4f2dfa0454f92ec93ae1a51d/ResourceUsage_09-01_09-30.csv)

SAB Downloaded: [resource_usage_09_01_09_30.csv.20191001131235.20191001131236.csv](/uploads/afa28b50cb33945fd2aca51d6d81c206/resource_usage_09_01_09_30.csv.20191001131235.20191001131236.csv)

As you can see all of the sub-accounts are missing from the downloaded file and also do not appear to be totaling into the parent account. I also ran the exception report and those sub-accounts are not listed, so I presume the upload process is finding them, but simply not adding them into the parent account?

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 30, 2019, 07:37

@kdjstudios @jredfield 

Thank you for the update.

> May we please have an estimate on adding the 2054 code into the 2051 code. I took a quick look at the 2054 code on the HR report but it does not have any value setup to be importing to it by the looks of it; same with 2052 and 2053. So may we please get the user and date when those codes were created, so we can confirm if the client requested any formula to go with the code.

| Activity Name | CDR Code | Cdr Field | Created AT              | Created By |
|---------------|----------|-----------|-------------------------|------------|
| SMS +         | 2054     | 2054      | 2019-10-28 22:04:17 UTC | jbrowne    |
| NCSM          | 2053     | 2053      | 2018-11-02 07:55:31 UTC | jbrowne    |
| NCSM          | 2052     | 2052      | 2018-11-01 21:44:10 UTC | jbrowne    |

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 30, 2019, 09:04

@amishra @smasih @jredfield 

Please respond to http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1619#note_50044. We need clarification on where the sub accounts usage is going and how it is being billed.

EDIT: If the sub accounts are not being totaled into the parent account correctly, we will need this fixed ASAP. please provide an estimate and start development on the fix right away.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 30, 2019, 13:01

Jo Ann Wrote:

> Do you think you will know something by tomorrow – I will do my billing on Friday morning.

I wrote:

> The support team is still looking into this. Once we have an update from them we will follow up accordingly. We have stressed the priority of this since your next billing is starting on Friday.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 31, 2019, 06:04

@kdjstudios @jredfield 

Thank you for the update.

As per my best understanding, we looked into this and our findings described below.

* We did not see any issue with parent & child account usage total.
* We looked randomly one child account `5387(Conviva Care Soluitons, LLC- San Antonio)` which parent account `1002(Conviva Care Solutions Parent Account)`.
* No resource id present on the usages file of the parent account (1002).
  * Child Account `5387` has resource-id `5387` which contains two rows in the Usages file 
and HR report shows the sum of the values of all rows.
   * Example -> 
      * Code -> 2027
        * (Maint:Time) / 60 -> 
          * first row value-> 218
          * second row value -> 235
          * Sum of all rows -> `218+235` = `453`
        * ((453)/60).ceil=> 8 
  

* `Parent Account Invoice` ->  [SPDZ-29667](/uploads/a64fb1bd63479640299f43777b941427/invoice.pdf)
* `Hr Report of the Child Account->` ![image](/uploads/0f2c2d476c2ba3de6c0cb9858fb020e3/image.png)
* `Upload Usage File:` [Uploaded Resource File](/uploads/3a54d0415c56a12c617b96efa5549326/ResourceUsage_09-01_09-30__1_.ods)  
* `Download Usage form Billing Cycle Page:` [Download Usage File](/uploads/505cece9a0e76e68ce26d9781ee9d2a0/download.ods) 

#### Let's see another example-> 
* Child account (`78 Peterson Specialty Care - Pulmonology`) which parent account (`5435 Peterson Specialty Care Group`)
* Both account resource id present on the Uploaded Usage file and downloaded usage file as well.
* No account activities on the parent account and **Use Parent Account Activities** are false.

* `Parent Account Invoice` -> [invoice_SPDZ-29846](/uploads/bf214aeb7129aadbc0a1c61dcc3220c7/invoice_SPDZ-29846.pdf) 
* `Hr Report of the Child Account->` ![image](/uploads/0f2c2d476c2ba3de6c0cb9858fb020e3/image.png).


##### In short, we did not find any issue with parent/child usages subtotal. It would be great if you have any specific account then let us know.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 31, 2019, 08:42

@amishra @smasih @jredfield 

Did you even read this ticket? We already outlined the issue and which account to use. It is even the subject line of this ticket?!!!!!

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 31, 2019, 09:07

@kdjstudios @jredfield 

Thank you for the update.

Sorry, I had completely forgotten I had just read your last comment.
let 's check the account(5195) as mentioned above and update you accordingly.
 
Thank you

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 31, 2019, 09:34

@kdjstudios @jredfield @smasih 

As per the current functionality of SAB:

* HR Report only shows its own account usage value which is upload from the usage file.

* But if you can see code (`2028`) on the calculation value column then the value shows the sum of all the rows values which are uploaded from the usages file.

![image](/uploads/ed9fa4d940a0666be98795792633ce8b/image.png)

Code Mapping: ![image](/uploads/33c5ab68a12aeda985c79f68bc397d80/image.png)

Usage File : -> ![Screenshot_from_2019-10-31_20-53-44](/uploads/140faa0034e6d37bcb481e48f0580c83/Screenshot_from_2019-10-31_20-53-44.png)

Hope this help

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 31, 2019, 10:50

@amishra @smasih @jredfield 

So here is a recap of all the issues I believe we have:

1. The downloaded usage file from SAB is not showing all rows. As shown in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1619#note_50044
2. The HR report does not display how many rows were summed. As shown in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1619#note_50378
3. The HR report appears to only sum certain metrics. It needs to sum all metrics. You can see the 7002 metric is not summed, but the 2028 appears to be. Sadly they are the exact same metric. As shown in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1619#note_50378
4. The summed usage is not then being billed properly. This is the main ticket.

Here are some of the files to better explain in one post all of the posts above.

[ResourceUsage_09-01_09-30_-_Highlighted_5195.xlsx](/uploads/f02a29df9f6afd2e509c31b5dbc96ec0/ResourceUsage_09-01_09-30_-_Highlighted_5195.xlsx) - This file is the filtered uploaded usage file which shows all the rows for the 5195 account and also highlights the SMP:Count metric and it's total of all rows. Which is not being calculated into the 2051 metric correctly.

[resource_usage_09_01_09_30.csv.20191001131235.20191001131236__4_.csv](/uploads/21d1e02e4558eb3ee773363ed6567477/resource_usage_09_01_09_30.csv.20191001131235.20191001131236__4_.csv) - This file shows the downloaded from SAB usage file, which is missing all of the sub account rows.

[ResourceUsage_09-01_09-30_-_Sub-accounts.xlsx](/uploads/5a9ef0a074764919d5274e014db699f4/ResourceUsage_09-01_09-30_-_Sub-accounts.xlsx)This file highlights all of the rows which have more then one with the same account number. Should be a list of all sub accounts.

### What we need to do, in the priority they need to be done. 

1. Fix the billing process to ensure all usage from child accounts is summed and being billed correctly.
2. Fix the HR report, to accurately display the usage.
3. Fix the download usage to accurately display all rows so it matches the uploaded usage file.
4. Find all of the accounts that this effected.
5. Determine how far back this has been going on for and hwich metrics were not being calculated correctly.
6. Determine the amount of missed revenue for each metric on each account on each billing cycle that was effected.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 1, 2019, 03:41

@kdjstudios @jredfield @tim.traylor 

#### Update

* Found the issue, it was in the SPDZ posting script.
* On 8^th June 2018 change for `2051` was made on the posting script.
* On 5^th June 2019 [+"SMP:Count"+] was added in the calculation of **`2051`**
* >> Issue: **Posting script has a section "CALCULATED FIELDS" and this code was not included in that section.**
  * **IMPORTANT:** However, the system was still billing the accounts as per the **USAGE FOUND IN FIRST ROW OF usage file** 
* `WE HAVE TESTED AND UPDATED THE POSTING SCRIPT (committed on git as well) ON LIVE SITE`. 
  * **Screen Shot From Human Readable of MERGE TEST SERVER** 
    ![image](/uploads/65b3d1864ff02f075eb0e958fbe28f44/image.png)
  * **Screen Shot From Invoice Draft of MERGE TEST SERVER**
    ![image](/uploads/b9831376fc8d0b552bcae9f7db538b45/image.png)  
#### Non-Billed Usage

* There are `634` accounts in SPDZ out of which `242` accounts have activity [+2051+]
  * There is a possibility that many accounts were terminated and created in this time span (8 June 2018 till now)
* Reported account [-5195-] should have been billed for usage 
  * Usage Units `4578` - Allowance `4000` = `578`
  * `578` x Excess Rate `0.35` = final [+202.3+]
  * Since 8^th June 2018 there has been 17 cycles 
* Now there are many factors that would have affected billing.
  * Because 2051 was introduced in June not sure how many accounts were using it.
  * System was always pulling usage from the first row **however,** it was not summing multiple rows.
  * Accounts could have been created or terminated in this time span.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 1, 2019, 04:42

assigned to @amishra

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 1, 2019, 06:09

#### Following are the codes which are present On the Site level (CDR Codes) but do not have a calculation defined in posting scripts

* 2043
* 2050
* 2052
* 2053
* 2054

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 1, 2019, 12:04

I wrote:


> The support team found the underlining miscalculation and has corrected it.  You should be able to run your billing and confirm that the SMP:Count metric is now calculating correctly into the 2051 Billing code for all subaccounts.  
> 
> Based upon their research so far they believe the issue appears to be that the SMP:Count metric was only including the parent account usage and none of the sub account's into the 2051 Billing code. So far this appears to only have effected the 2051 Billing code in regards to the SMP:Count Metric.  They believe this miscalculation was introduced when the SMP:Count metric was introduced in the the 2051 billing code back in July 2019. We are having them continue their research to provide further details and we will keep you updated. 
> 
> Please let us know if you have any questions or concerns.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 5, 2019, 13:29

@amishra @smasih @jredfield 

May you please confirm where we are at on completing these steps below. I have added what I believe is the status of each. Please provide an estimate on each task. 

1. Fix the billing process to ensure all usage from child accounts is summed and being billed correctly. **[Completed, waiting confirmation on next billing cycle]**
1. Fix the HR report, to accurately display the usage. **[Completed, no issue with HR was with posting script.]**
1. Fix the download usage to accurately display all rows so it matches the uploaded usage file. **[Not Completed]**
1. Find all of the accounts that this effected. **[Not Completed]**
1. Determine how far back this has been going on for and which metrics were not being calculated correctly. **[Completed]**
1. Determine the amount of missed revenue for each metric on each account on each billing cycle that was effected.  **[Not Completed]**

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 8, 2019, 05:23

@kdjstudios @jredfield 


Thank you for the update.



> Fix the download usage to accurately display all rows so it matches the uploaded usage file.

* It'll take more time because we need to change the posting script and we found one another ticket #1465 related with the same. So we will update there.

> Find all of the accounts that this effected.

* We can find the account as per the current status of the account and its activities but we **can't** find the exact account that was affected since 8^th June 2018 because the user has the ability to activate & deactivate the fee anytime  `[Approx 2-4]`


> Determine the amount of missed revenue for each metric on each account on each billing cycle that was effected.

* We **can not** determine the exact amount. What we can do is multiple with billing cycle count with the average amount missed each month since 8^June 18. `[Approx 3-5]`


## Developer Estimate
|Size|Lower Time Bound|Variability|Upper Time Bound|Testing|Time for Testing| Time for Dev & Testing|
|-|-|-|-|-|-|-|
|Medium-2|5|Aware of Concepts, implementation tricky|4|NA|0.0|9|

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 8, 2019, 08:33

@amishra @smasih @jredfield 

If I am reading this correctly there is still the issue with the posting script/billing process not calculating correctly?! THIS HAS TO BE FIXED IMMEDIATELY! Marking back to development just for that. This needs to be released ASAP, as we can not have another billing cycle be messed up.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 9, 2019, 04:44

@kdjstudios @jredfield 

Thank you for the update.

Let me clear here:

* The main issue of this ticket was not summed child accounts usages in the parent account. **This issue has completed & already on the production server. now, All child accounts usages summed with parent account usage.**
* The CDR file (SAB Download file) needs to match the uploaded file. Already has ticket #1465 and we will fix this issue under that ticket.

* Rest of the remaining task for this ticket is points 3, 4 and 6 mentioned above on your  [comment](http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1619#note_50641) 

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 11, 2019, 10:39

@amishra @smasih @jredfield 


Looks like the below was completed too.

1. Fix the download usage to accurately display all rows so it matches the uploaded usage file. **[Completed, waiting release on #1465 ]**

We are still needing estimates and to have these completed so we can inform the client of their missed revenue. Please let us know if there is anything needed from us for you to complete these. 

1. Find all of the accounts that this effected. **[Not Completed]**
1. Determine the amount of missed revenue for each metric on each account on each billing cycle that was effected. **[Not Completed]**

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 13, 2019, 04:25

@kdjstudios @jredfield 

## Developer Estimate
|Size|Lower Time Bound|Variability|Upper Time Bound|Testing|Time for Testing| Time for Dev & Testing|
|-|-|-|-|-|-|-|
|Medium-2|6|Worked with Concepts, implementation ok|9|NA|0.0|9|

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 13, 2019, 14:41

@jredfield 

This is a bit more time then I thought it would take to find out how much revenue was lost/not billed for. We may want to reach out to Jo Ann, first to confirm if how badly they would need this information.

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Nov 19, 2019, 13:39

@kdjstudios @amishra Holding off on approval until discussion with Jo Ann.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 22, 2019, 10:50

I wrote:

> We are just following up as it came to our attention that due to the SMP:Count metric miscalculation that you have been billing your clients less then what they should have. Our support team believes they may be able to look back on the accounts that were affected and possibly determine how much usage/revenue may have been missed. It would take the team a few days to find this, so we did want to reach out and see if you were interested in that information?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 22, 2019, 12:37

Jo Ann wrote:

> No thanks. I will not go back and bill them as I will probably lose most of those customers if I tried. Is there anything else I had to do to make it work on my end

I wrote:

> There should be nothing on your end to do. I would simply ask that when you are doing your next billing once you have done the apply step to let us know. That way I may have the support team just double check everything worked as expected before you finalize and send out invoices. I am sure it will work perfectly, but I certainly do not want there to be any unforeseen concerns arise.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 27, 2019, 10:36

@smasih @amishra 

I was just reviewing the release of #1465 and just want to confirm that the update released in this ticket is only going to effect future imports and billing cycles correct? I presume so, because, the updated HR report does not show that there were multiple rows or the total of the child import values in the SMP metrics on the last billing cycle for account 5195. However, it does appear that the calculated field 2021 is calculating the total correctly, which is a bit puzzling. 

![image](/uploads/0c2763a2e25e3e5075df6655b930d0bd/image.png)

![image](/uploads/a2dfb77fdb7be47a3f47f153cc2c16f5/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Nov 27, 2019, 22:44

@kdjstudios @jredfield 

> I was just reviewing the release of #1465 and just want to confirm that the update released in this ticket is only going to effect future imports and billing cycles correct?

Yes, from now on it will store all the uploaded usage rows in different cdr records. Previously, it was clubbing them into the single cdr record.

> I presume so, because, the updated HR report does not show that there were multiple rows or the total of the child import values in the SMP metrics on the last billing cycle for account 5195. However, it does appear that the calculated field 2021 is calculating the total correctly, which is a bit puzzling.

We were storing the values for SMP metrics in the cdr record. The only issue was with the HR report UI/view.
After we fixed the HR Report view in the release it will show all the correct value for the 2051 code, as we already had the records saved for the SMP metrics in the cdr record for the previous billing cycles.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 29, 2019, 10:20

@ranand 

Thank you. We should have feedback once Speed Ez completes there next billing if everything is working.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 4, 2019, 12:42

Everything looks great. closing ticket.
